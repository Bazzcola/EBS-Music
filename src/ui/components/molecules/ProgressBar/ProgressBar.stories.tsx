import React from 'react';
import { ProgressBar } from 'ui/components/molecules/ProgressBar/ProgressBar';

export default {
  title: 'Molecules/ProgressBar',
  component: ProgressBar
};

export const ProgresBar1 = () => {
  return <ProgressBar />;
};
