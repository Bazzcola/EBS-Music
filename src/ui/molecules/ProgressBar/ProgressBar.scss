.progress_bar {
    max-width: 720px;
    margin: 0 auto;
    text-align: center;
    height: 50px;
    box-sizing: border-box;

    .slider_track_input {
        width: 500px;
        margin: 0px;
    }

    .current_Time {
        margin-right: 20px;
        color: #92929d;
        font-size: 10px;
    }

    .total_Time {
        margin-left: 20px;
        color: #92929d;
        font-size: 10px;
    }
}

.bar_wrapper {
    padding-top: 10px;
}

.song_text {
    display: flex;
    justify-content: flex-start;
    margin-left: 112px;
}

.song_group {
    font-family: Deezer, Arial, sans-serif;
    font-weight: 400;
    margin-right: 5px;
}

.song_name {
    font-weight: 400;
    font-family: Open Sans, Arial, sans-serif;
    margin-left: 5px;
}