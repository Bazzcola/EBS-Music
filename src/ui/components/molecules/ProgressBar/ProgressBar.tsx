import React, { useRef, useContext, useEffect, useState } from 'react';
import { AudioContext, Tracks, song } from 'ui/context/audioContext';
import ReactTooltip from 'react-tooltip';
import 'ui/components/molecules/ProgressBar/ProgressBar.scss';

export const ProgressBar = () => {
  const {
    currentTimeSecond,
    setCurrentTimeSecond,
    durationTime,
    setDurationTime,
    clickedTime,
    setClickedTime,
    playing,
    setPlaying,
    volume,
    audioFiles,
    setAudioFiles,
    counter,
    setCounter,
    shuffle,
    repeatOne,
    repeatAll,
    profile
  } = useContext(AudioContext);
  const [barTooltip, setBarTooltip] = useState<number>(0);
  const [files, setFiles] = useState<string[]>(['']);
  const [random, setRandom] = useState<number>(1);
  const [currentData, setCurrentData] = useState<song>();
  const [prevSong, setPrevSong] = useState<number>();
  const barDiv = useRef<any>(null);
  const audio = useRef<any>(null);
  const cur = audio.current;

  useEffect(() => {
    if(profile) {
      (async function anyNameFunction() {
        await cur.load();
        await cur.pause();
      })();
    }
  },[profile])

  useEffect(() => {
    setFiles(Tracks.map((item: song) => item.src));
  }, []);

  useEffect(() => {
    Tracks.filter((item: song) =>
      item.src === audioFiles ? setCurrentData(item) : false
    );
  },[currentTimeSecond, audioFiles]);

  useEffect(() => {
    const cur = audio.current;
    playing ? (async function anyNameFunction() {
      await cur.play();
    })() : (async function anyNameFunction() {
      await cur.load();
      await cur.pause();
    })();
  },[playing])

  useEffect(() => {
    const cur = audio.current;
    cur.ontimeupdate = () => {
      let seconds = cur.currentTime;
      setCurrentTimeSecond(seconds.toFixed());
    };
    cur.onloadedmetadata = () => {
      let allSeconds = cur.duration;
      setDurationTime(allSeconds.toFixed());
    };

    cur.volume = volume;

    if (barTooltip < 0) {
      setBarTooltip(0);
    }
    if (clickedTime && clickedTime !== currentTimeSecond) {
      cur.currentTime = clickedTime;
      setClickedTime(0);
    }
    // eslint-disable-next-line
  }, [currentTimeSecond, clickedTime, playing, barTooltip, volume]);

  let keys: any = Object.keys(files);

  const Next = async () => {
    if (currentTimeSecond > 10 && currentTimeSecond === durationTime) {
      console.log('закончилась песня');
      if (repeatAll && shuffle === false) {
        if (counter === files.length - 1) {
          await cur.play();
          setCounter(0);
          setAudioFiles(files[keys[0]]);
        } else {
          await cur.play();
          setCounter((count) => count + 1);
          setAudioFiles(files[keys[counter + 1]]);
          setPrevSong(counter + 1);
        }
      } else if(repeatAll && shuffle === true) {
        if(random === prevSong) {
          await cur.play();
          setRandom(Math.floor(Math.random() * 8))
          setCounter(random)
          setAudioFiles(files[keys[random + 1]])
        } else {
          await cur.play();
          setRandom(Math.floor(Math.random() * 9))
          setCounter(random)
          setAudioFiles(files[keys[random]])
          setPrevSong(random);
        }
        
      } else {
        await cur.load();
        await cur.pause();
        setPlaying(false);
        console.log(playing + ' ошибка')
      }
      
    }
  };

  const calcClickedTime = (e: { pageX: number }) => {
    const clickPositionInPage = e.pageX;
    const bar = barDiv.current;
    const barStart = bar.getBoundingClientRect().left + window.scrollX;
    const barWidth = bar.offsetWidth;
    const clickPositionInBar = clickPositionInPage - barStart;
    const timePerPixel = durationTime / barWidth;
    const result = timePerPixel * clickPositionInBar;
    return result.toFixed();
  };

  const onTimeUpdate = (time: any) => setClickedTime(time);

  const handleTimeDrag = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    e.persist();
    onTimeUpdate(calcClickedTime(e));
    const onMove = (eMove: { pageX: number }) =>
      onTimeUpdate(calcClickedTime(eMove));

    document.addEventListener('mousemove', onMove);
    document.addEventListener('mouseup', () => {
      document.removeEventListener('mousemove', onMove);
    });
  };

  const barTooltipLine = (time: any) => setBarTooltip(time);

  const getTooltipTime = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    e.persist();
    const onMoveTooltip = (eMove: { pageX: number }) =>
      barTooltipLine(calcClickedTime(eMove));

    document.addEventListener('mousemove', onMoveTooltip);
    document.addEventListener('mouseup', () => {
      document.removeEventListener('mousemove', onMoveTooltip);
    });
  };

  const formatTooltiptime = (secs: number) => {
    var hr = Math.floor(secs / 3600);
    var min = Math.floor((secs - hr * 3600) / 60);
    var sec = Math.floor(secs - hr * 3600 - min * 60);

    if (min < 10) {
      min = 0 + min;
    }
    if (sec < 10) {
      sec = 0 + sec;
    }
    // eslint-disable-next-line
    return min < 10 && sec < 10 ? min + ':' + '0' + sec : min + ':' + sec;
  };
  const formatSecondsAsTime = (secs: number) => {
    var hr = Math.floor(secs / 3600);
    var min = Math.floor((secs - hr * 3600) / 60);
    var sec = Math.floor(secs - hr * 3600 - min * 60);

    // eslint-disable-next-line
    return min < 10 && sec < 10 ? min + ':' + '0' + sec : min + ':' + sec;
  };
  const formatSecondsAllTime = (secs: number) => {
    var hr = Math.floor(secs / 3600);
    var min = Math.floor((secs - hr * 3600) / 60);
    var sec = Math.floor(secs - hr * 3600 - min * 60);
    // eslint-disable-next-line
    return min < 10 && sec < 10 ? min + ':' + '0' + sec : min + ':' + sec;
  };

  const curPercentage = (currentTimeSecond / durationTime) * 100;

  return (
    <div className="progress_bar">
      <audio
        loop={repeatOne}
        autoPlay
        preload="metadata"
        id="audio"
        ref={audio}
        className="audio_source"
        src={audioFiles}
        onEnded={Next}
      ></audio>
      <div className="song_text">
        <span className="song_group">{currentData?.title} {currentData ? '•' : ''}</span><span className="song_name">{currentData?.name}</span>
      </div>
      <div className="bar">
        <span className="bar__time">
          {formatSecondsAsTime(currentTimeSecond)}
        </span>
        <div
          data-tip=""
          data-for="bar_time"
          className="bar__progress"
          ref={barDiv}
          style={{
            background: `linear-gradient(to right, #6fd44a ${curPercentage}%, grey 0)`
          }}
          onMouseDown={(e) => handleTimeDrag(e)}
          onMouseMove={(e) => getTooltipTime(e)}
        >
          <span
            className="bar__progress__knob"
            style={{ left: `${curPercentage - 2}%` }}
          />
          <ReactTooltip id="bar_time">
            {formatTooltiptime(barTooltip)}
          </ReactTooltip>
        </div>

        <span className="bar__time">
          {playing || durationTime > 1
            ? formatSecondsAllTime(durationTime)
            : '0:00'}
        </span>
      </div>
    </div>
  );
};
