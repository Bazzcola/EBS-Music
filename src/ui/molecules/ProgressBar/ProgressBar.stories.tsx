import React from 'react';
import { ProgressBar } from 'ui/molecules/ProgressBar/ProgressBar';

export default {
  title: 'Molecules/ProgressBar',
  component: ProgressBar
};

export const ProgresBar1 = () => {
  return <ProgressBar />;
};
