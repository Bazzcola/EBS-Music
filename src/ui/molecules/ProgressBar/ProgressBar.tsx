import React, { useRef, useContext, useEffect, useState } from 'react';
import { AudioContext } from 'ui/context/audioContext';
import 'ui/molecules/ProgressBar/ProgressBar.scss';

export const ProgressBar = () => {
  const { currentTimeSecond, setCurrentTimeSecond } = useContext(AudioContext);
  const { durationTime, setDurationTime } = useContext(AudioContext);
  const { clickedTime, setClickedTime } = useContext(AudioContext);
  const [playing, setPlaying] = useState(false);
  const audio = useRef(null);

  useEffect(() => {
    const cur: any = audio.current;
    cur.ontimeupdate = () => {
      let seconds = cur.currentTime;
      setCurrentTimeSecond(seconds.toFixed());
    };
    cur.onloadedmetadata = () => {
      let allSeconds = cur.duration;
      setDurationTime(allSeconds.toFixed());
    };
    if (clickedTime && clickedTime !== currentTimeSecond) {
      cur.currentTime = clickedTime;
      setClickedTime(0);
    }
  }, [currentTimeSecond, clickedTime]);

  // console.log(durationTime);
  // console.log(currentTimeSecond);
  // console.log(clickedTime);

  // useEffect(() => {
  //   const cur: any = audio.current;

  //   const setAudioData = () => {
  //     setDurationTime(cur.duration);
  //     setCurrentTimeSecond(cur.currentTime);
  //   }

  //   const setAudioTime = () => setCurrentTimeSecond(cur.currentTime);

  //   cur.addEventListener("loadeddata", setAudioData);

  //   cur.addEventListener("timeupdate", setAudioTime);

  //   playing ? cur.play() : cur.pause();

  //   if (clickedTime && clickedTime !== currentTimeSecond) {
  //     cur.currentTime = clickedTime;
  //     setClickedTime(0);
  //   } 

  //   return () => {
  //     cur.removeEventListener("loadeddata", setAudioData);
  //     cur.removeEventListener("timeupdate", setAudioTime);
  //   }
  // },[]);
 
  const calcClickedTime = (e: { pageX: number }) => {
    const clickPositionInPage = e.pageX;
    const bar: any = document.querySelector('.bar__progress');
    const barStart = bar.getBoundingClientRect().left + window.scrollX;
    const barWidth = bar.offsetWidth;
    const clickPositionInBar = clickPositionInPage - barStart;
    const timePerPixel = durationTime / barWidth;
    const result = timePerPixel * clickPositionInBar;
    console.log(result.toFixed())
    setClickedTime(result);
    // return result.toFixed();
  };
 
  // const handleTimeDrag = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
  //   e.persist();
  //   const onTimeUpdate = (eMove: string) => calcClickedTime(e);

  //   const updateTimeOnMove = (eMove: { pageX: number; }) => {
  //     onTimeUpdate(calcClickedTime(eMove));
  //   };

  //   document.addEventListener('mousemove', updateTimeOnMove);

  //   document.addEventListener('mouseup', () => {
  //     document.removeEventListener('mousemove', updateTimeOnMove);
  //   });
  // };

  const Play = () => {
    const audioPlay: any = audio.current;
    audioPlay.play();
  };
  const Stop = () => {
    const audioPlay: any = audio.current;
    audioPlay.pause();
  };
  const formatSecondsAsTime = (secs: number) => {
    var hr = Math.floor(secs / 3600);
    var min = Math.floor((secs - hr * 3600) / 60);
    var sec = Math.floor(secs - hr * 3600 - min * 60);

    if (min < 10) {
      min = 0 + min;
    }
    if (sec < 10) {
      sec = 0 + sec;
    }

    return min < 10 && sec < 10 ? min + ':' + '0' + sec : min + ':' + sec;
  };
  const formatSecondsAllTime = (secs: number) => {
    var hr = Math.floor(secs / 3600);
    var min = Math.floor((secs - hr * 3600) / 60);
    var sec = Math.floor(secs - hr * 3600 - min * 60);

    if (min < 10) {
      min = 0 + min;
    }
    if (sec < 10) {
      sec = 0 + sec;
    }

    return min + ':' + sec;
  };

  const curPercentage = (currentTimeSecond / durationTime) * 100;

  return (
    <div className="progress_bar">
      <audio controls preload="metadata" id="audio" ref={audio}>
        <source
          src="https://www.soundhelix.com/examples/mp3/SoundHelix-Song-1.mp3"
          type="audio/mpeg"
        />
      </audio>
      <div className="song_text">
        <span className="song_group">ALABAMA</span>â€¢
        <span className="song_name">Kyok</span>
      </div>
      <div className="bar">
        <span className="bar__time">
          {formatSecondsAsTime(currentTimeSecond)}
        </span>
        <div
          className="bar__progress"
          style={{
            background: `linear-gradient(to right, #6fd44a ${curPercentage}%, grey 0)`
          }}
          onMouseDown={(e) => calcClickedTime(e)}
        >
          <span
            className="bar__progress__knob"
            style={{ left: `${curPercentage - 2}%` }}
          />
        </div>

        <span className="bar__time">{formatSecondsAllTime(durationTime)}</span>
      </div>
      <button onClick={Play}>PLay</button>
      <button onClick={Stop}>Stop</button>
    </div>
  );
};
