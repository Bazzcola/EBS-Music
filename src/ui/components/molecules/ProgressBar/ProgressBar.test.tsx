import React from 'react';
import { render } from '@testing-library/react';
import { ProgressBar } from './ProgressBar';

it('renders ProgressBar correctly', () => {
  const { asFragment } = render(<ProgressBar />);
  expect(asFragment()).toMatchSnapshot();
});
