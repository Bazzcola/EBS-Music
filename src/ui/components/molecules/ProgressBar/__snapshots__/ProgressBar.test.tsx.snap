// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ProgressBar correctly 1`] = `
<DocumentFragment>
  <div
    class="progress_bar"
  >
    <audio
      autoplay=""
      class="audio_source"
      id="audio"
      preload="metadata"
      src=""
    />
    <div
      class="song_text"
    >
      <span
        class="song_group"
      >
         
      </span>
      <span
        class="song_name"
      />
    </div>
    <div
      class="bar"
    >
      <span
        class="bar__time"
      >
        0:00
      </span>
      <div
        class="bar__progress"
        currentitem="false"
        data-for="bar_time"
        data-tip=""
      >
        <span
          class="bar__progress__knob"
          style="left: -2%;"
        />
        <div
          class="__react_component_tooltip ta14425b5-6748-4bbf-b9db-7d54819d1daa place-top type-dark"
          data-id="tooltip"
          id="bar_time"
        >
          <style>
            
  	.ta14425b5-6748-4bbf-b9db-7d54819d1daa {
	    color: #fff;
	    background: #222;
	    border: 1px solid transparent;
  	}

  	.ta14425b5-6748-4bbf-b9db-7d54819d1daa.place-top {
        margin-top: -10px;
    }
    .ta14425b5-6748-4bbf-b9db-7d54819d1daa.place-top::before {
        border-top: 8px solid transparent;
    }
    .ta14425b5-6748-4bbf-b9db-7d54819d1daa.place-top::after {
        border-left: 8px solid transparent;
        border-right: 8px solid transparent;
        bottom: -6px;
        left: 50%;
        margin-left: -8px;
        border-top-color: #222;
        border-top-style: solid;
        border-top-width: 6px;
    }

    .ta14425b5-6748-4bbf-b9db-7d54819d1daa.place-bottom {
        margin-top: 10px;
    }
    .ta14425b5-6748-4bbf-b9db-7d54819d1daa.place-bottom::before {
        border-bottom: 8px solid transparent;
    }
    .ta14425b5-6748-4bbf-b9db-7d54819d1daa.place-bottom::after {
        border-left: 8px solid transparent;
        border-right: 8px solid transparent;
        top: -6px;
        left: 50%;
        margin-left: -8px;
        border-bottom-color: #222;
        border-bottom-style: solid;
        border-bottom-width: 6px;
    }

    .ta14425b5-6748-4bbf-b9db-7d54819d1daa.place-left {
        margin-left: -10px;
    }
    .ta14425b5-6748-4bbf-b9db-7d54819d1daa.place-left::before {
        border-left: 8px solid transparent;
    }
    .ta14425b5-6748-4bbf-b9db-7d54819d1daa.place-left::after {
        border-top: 5px solid transparent;
        border-bottom: 5px solid transparent;
        right: -6px;
        top: 50%;
        margin-top: -4px;
        border-left-color: #222;
        border-left-style: solid;
        border-left-width: 6px;
    }

    .ta14425b5-6748-4bbf-b9db-7d54819d1daa.place-right {
        margin-left: 10px;
    }
    .ta14425b5-6748-4bbf-b9db-7d54819d1daa.place-right::before {
        border-right: 8px solid transparent;
    }
    .ta14425b5-6748-4bbf-b9db-7d54819d1daa.place-right::after {
        border-top: 5px solid transparent;
        border-bottom: 5px solid transparent;
        left: -6px;
        top: 50%;
        margin-top: -4px;
        border-right-color: #222;
        border-right-style: solid;
        border-right-width: 6px;
    }
  
          </style>
          0:00
        </div>
      </div>
      <span
        class="bar__time"
      >
        0:00
      </span>
    </div>
  </div>
</DocumentFragment>
`;
